File: sdk/python/examples/quickstart.py
"""Aegis SDK quickstart — and the v0.1 end-to-end smoke test.

This is both the canonical "drop two lines in your app" example and a real
round-trip check: it runs the actual SDK against a live Aegis backend and a real
LLM provider (Groq's free tier), then reads the traces back out to prove the whole
path works:

    aegis.instrument()  ->  real LLM call  ->  ring buffer  ->  flush thread
        ->  POST /v1/traces/batch  ->  ClickHouse  ->  GET /v1/traces

Prerequisites:
  * Aegis stack running:   docker compose up -d
  * A seeded API key:      docker compose exec api python -m scripts.seed_demo_tenant
  * Groq free-tier key:    already in the repo .env as GROQ_API_KEY

Run (PowerShell):
    $env:AEGIS_API_KEY="aegis_dev_..."; $env:GROQ_API_KEY="gsk_..."
    .venv\\Scripts\\python.exe examples\\quickstart.py
"""

from __future__ import annotations

import os
import sys

import httpx
from openai import OpenAI

from aegis_sdk import Aegis

# Windows consoles default to cp1252 and choke on the ✓/✗ this script prints.
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

AEGIS_BASE_URL = os.environ.get("AEGIS_BASE_URL", "http://localhost:8000")
AEGIS_API_KEY = os.environ.get("AEGIS_API_KEY", "")
GROQ_API_KEY = os.environ.get("GROQ_API_KEY", "")
GROQ_BASE_URL = "https://api.groq.com/openai/v1"
GROQ_MODEL = "llama-3.3-70b-versatile"  # free tier; verify at console.groq.com/docs/deprecations


def _trace_count(http: httpx.Client) -> int:
    """Total traces visible to this API key, via the read API."""
    resp = http.get("/v1/traces", params={"limit": 1})
    resp.raise_for_status()
    return int(resp.json()["meta"]["total"])


def main() -> int:
    if not AEGIS_API_KEY.startswith("aegis_"):
        sys.exit("AEGIS_API_KEY missing/malformed — seed a tenant and export it first.")
    if not GROQ_API_KEY.startswith("gsk_"):
        sys.exit("GROQ_API_KEY missing/malformed — expected `gsk_...` (see repo .env).")

    read_api = httpx.Client(
        base_url=AEGIS_BASE_URL,
        headers={"Authorization": f"Bearer {AEGIS_API_KEY}"},
        timeout=10.0,
    )
    before = _trace_count(read_api)
    print(f">>> traces before: {before}")

    # --- the two lines a real app adds -------------------------------------
    aegis = Aegis(api_key=AEGIS_API_KEY, base_url=AEGIS_BASE_URL, project="sdk-quickstart")
    client = aegis.instrument(OpenAI(api_key=GROQ_API_KEY, base_url=GROQ_BASE_URL))
    # -----------------------------------------------------------------------

    # 1. a plain non-streaming call
    print(">>> non-streaming call...")
    resp = client.chat.completions.create(
        model=GROQ_MODEL,
        messages=[{"role": "user", "content": "In one sentence, what is observability?"}],
        aegis_metadata={"example": "quickstart", "kind": "non-stream"},
    )
    print(f"    {resp.choices[0].message.content[:80]}...")

    # 2. a streaming call — chunks pass straight through; the trace is finalized
    #    after the stream completes, with full token + cost rollup
    print(">>> streaming call...")
    stream = client.chat.completions.create(
        model=GROQ_MODEL,
        messages=[{"role": "user", "content": "Name three LLM failure modes, very briefly."}],
        stream=True,
        aegis_metadata={"example": "quickstart", "kind": "stream"},
    )
    streamed = "".join(
        chunk.choices[0].delta.content
        for chunk in stream
        if chunk.choices and chunk.choices[0].delta.content
    )
    print(f"    {streamed[:80]}...")

    # flush + clean shutdown — guarantees both traces are shipped before we read back
    aegis.close()

    after = _trace_count(read_api)
    print(f">>> traces after:  {after}  (delta {after - before:+d})")

    # newest trace should be our streamed call: groq model, real tokens, real completion
    latest = read_api.get("/v1/traces", params={"limit": 1}).json()["traces"][0]
    print(f">>> latest trace:  model={latest['model']} "
          f"tokens={latest['tokens_in']}/{latest['tokens_out']} "
          f"cost=${latest['cost_usd']:.6f} latency={latest['latency_ms']}ms")

    read_api.close()

    ok = (
        after - before == 2
        and latest["model"] == f"groq/{GROQ_MODEL}"
        and latest["tokens_out"] > 0
    )
    print("\nE2E smoke:", "PASS ✓" if ok else "FAIL ✗")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: sdk/typescript/examples/quickstart.ts
// Aegis SDK quickstart — and the v0.2 TypeScript end-to-end smoke test.
//
// Same shape as sdk/python/examples/quickstart.py: instrument a real OpenAI
// client pointed at Groq, make one non-streaming and one streaming call, close
// the SDK to flush, then read the traces back via the read API to prove the
// full path:
//
//   aegis.instrument()  →  real LLM call  →  ring buffer  →  setInterval flush
//        →  POST /v1/traces/batch  →  ClickHouse  →  GET /v1/traces
//
// Prerequisites:
//   * Aegis stack up:           docker compose up -d
//   * A seeded API key:         docker compose exec api python -m scripts.seed_demo_tenant
//   * GROQ_API_KEY in env:      already in the repo .env
//
// Run (PowerShell):
//   $env:AEGIS_API_KEY="aegis_dev_..."; $env:GROQ_API_KEY="gsk_..."
//   npx tsx examples/quickstart.ts

import OpenAI from "openai";

import { Aegis } from "../src/index.js";

const AEGIS_BASE_URL = process.env.AEGIS_BASE_URL ?? "http://localhost:8000";
const AEGIS_API_KEY = process.env.AEGIS_API_KEY ?? "";
const GROQ_API_KEY = process.env.GROQ_API_KEY ?? "";
const GROQ_BASE_URL = "https://api.groq.com/openai/v1";
// Free tier; verify periodically at console.groq.com/docs/deprecations.
const GROQ_MODEL = "llama-3.3-70b-versatile";

async function traceCount(headers: Record<string, string>): Promise<number> {
  const resp = await fetch(`${AEGIS_BASE_URL}/v1/traces?limit=1`, { headers });
  if (!resp.ok) throw new Error(`read API ${resp.status}`);
  const body = (await resp.json()) as { meta: { total: number } };
  return body.meta.total;
}

async function main(): Promise<number> {
  if (!AEGIS_API_KEY.startsWith("aegis_")) {
    console.error("AEGIS_API_KEY missing/malformed — seed a tenant and export it first.");
    return 1;
  }
  if (!GROQ_API_KEY.startsWith("gsk_")) {
    console.error("GROQ_API_KEY missing/malformed — expected `gsk_...` (see repo .env).");
    return 1;
  }

  const readHeaders = { Authorization: `Bearer ${AEGIS_API_KEY}` };
  const before = await traceCount(readHeaders);
  console.log(`>>> traces before: ${before}`);

  // --- the two lines a real app adds -------------------------------------
  const aegis = new Aegis({
    apiKey: AEGIS_API_KEY,
    baseUrl: AEGIS_BASE_URL,
    project: "ts-sdk-quickstart",
  });
  const client = aegis.instrument(
    new OpenAI({ apiKey: GROQ_API_KEY, baseURL: GROQ_BASE_URL }),
  );
  // -----------------------------------------------------------------------

  // 1. Non-streaming call
  console.log(">>> non-streaming call...");
  const resp = await client.chat.completions.create({
    model: GROQ_MODEL,
    messages: [{ role: "user", content: "In one sentence, what is observability?" }],
    // `aegis_metadata` is popped by instrument(); never reaches Groq.
    aegis_metadata: { example: "quickstart", kind: "non-stream" },
  } as unknown as Parameters<typeof client.chat.completions.create>[0]);
  // `resp` here is the ChatCompletion (no stream).
  console.log(`    ${("choices" in resp ? resp.choices[0]?.message?.content ?? "" : "").slice(0, 80)}...`);

  // 2. Streaming call — chunks pass straight through; trace finalized on
  //    stream completion with full token + cost rollup.
  console.log(">>> streaming call...");
  const stream = await client.chat.completions.create({
    model: GROQ_MODEL,
    messages: [{ role: "user", content: "Name three LLM failure modes, very briefly." }],
    stream: true,
    aegis_metadata: { example: "quickstart", kind: "stream" },
  } as unknown as Parameters<typeof client.chat.completions.create>[0]);

  const parts: string[] = [];
  // The wrapper is an async iterable of chunks.
  for await (const chunk of stream as unknown as AsyncIterable<{
    choices: Array<{ delta: { content?: string | null } }>;
  }>) {
    const piece = chunk.choices[0]?.delta?.content;
    if (piece) parts.push(piece);
  }
  console.log(`    ${parts.join("").slice(0, 80)}...`);

  // Flush + clean shutdown — guarantees both traces ship before we read back.
  await aegis.close();

  const after = await traceCount(readHeaders);
  console.log(`>>> traces after:  ${after}  (delta ${after - before >= 0 ? "+" : ""}${after - before})`);

  // Newest trace should be our streamed call: groq model, real tokens.
  const listResp = await fetch(`${AEGIS_BASE_URL}/v1/traces?limit=1`, { headers: readHeaders });
  const list = (await listResp.json()) as {
    traces: Array<{ model: string; tokens_in: number; tokens_out: number; cost_usd: number; latency_ms: number }>;
  };
  const latest = list.traces[0]!;
  console.log(
    `>>> latest trace:  model=${latest.model} tokens=${latest.tokens_in}/${latest.tokens_out} ` +
      `cost=$${latest.cost_usd.toFixed(6)} latency=${latest.latency_ms}ms`,
  );

  const ok =
    after - before === 2 &&
    latest.model === `groq/${GROQ_MODEL}` &&
    latest.tokens_out > 0;
  console.log(`\nE2E smoke: ${ok ? "PASS" : "FAIL"}`);
  return ok ? 0 : 1;
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
